File: algorithms/code/leetcode/lc384_shuffle_an_array/LC384ShuffleAnArray.java
package leetcode.lc384_shuffle_an_array;

import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Random;

/**
 * https://leetcode.com/problems/shuffle-an-array/
 * #medium
 */
public final class LC384ShuffleAnArray {
    public static class Solution {
        private static final Random RANDOM = new SecureRandom();
        private final int[] nums;

        Solution(final int[] numsValue) {
            nums = numsValue;
        }

        public final int[] reset() {
            return Arrays.copyOf(nums, nums.length);
        }

        public final int[] shuffle() {
            if (nums == null) {
                return null;
            }
            int[] result = nums.clone();
            for (int i = nums.length - 1; i > 0; i--) {
                int j = RANDOM.nextInt(i + 1);
                int exchange = result[i];
                result[i] = result[j];
                result[j] = exchange;
            }
            return result;
        }
    }
}
package leetcode.lc384_shuffle_an_array;

import org.junit.jupiter.api.Test;

import leetcode.lc384_shuffle_an_array.LC384ShuffleAnArray.Solution;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

public final class LC384ShuffleAnArrayTests {
    private void test(final int[] nums) {
        Solution solution = new Solution(nums.clone());
        int[] result = solution.shuffle();
        int[] reset = solution.reset();
        assertArrayEquals(nums, reset);
        Arrays.sort(nums);
        Arrays.sort(result);
        assertArrayEquals(nums, result);
    }

    @Test
    public void testExample() throws Exception {
        int[] nums = { 1, 2, 3 };
        test(nums);
        // Should use Chi-squared test
    }

    @Test
    public void testNothing() throws Exception {
        assertNotNull(new LC384ShuffleAnArray());
        Solution solution = new Solution(null);
        assertNull(solution.shuffle());
    }
}

File: algorithms/code/leetcode/lc384_shuffle_an_array/lc384_shuffle_an_array.py
# https://leetcode.com/problems/shuffle-an-array/

import random
import unittest


class Solution:
    def __init__(self, nums: list[int]) -> None:
        self.__nums = nums

    def reset(self) -> list[int]:
        return self.__nums.copy()

    def shuffle(self) -> list[int]:
        result = self.__nums.copy()
        for i in range(len(self.__nums) - 1, 0, -1):
            j = random.SystemRandom().randint(0, i)
            result[i], result[j] = result[j], result[i]
        return result


class TestCode(unittest.TestCase):
    def __test(self, nums: list[int]) -> None:
        solution = Solution(nums.copy())
        result = solution.shuffle()
        reset = solution.reset()
        assert nums == reset
        nums.sort()
        result.sort()
        assert nums == result

    def test_example(self) -> None:
        nums = [1, 2, 3]
        self.__test(nums)
        # Should use Chi-squared test

    def test_nothing(self) -> None:
        solution = Solution([])
        assert not solution.shuffle()
